### PR TITLE
Add support for exponentiation

### PR DESCRIPTION
**Pull request by tohrnii**
Addressing #13.
Adds exponentiation to entire pipeline.

- [x] scanning
- [x] parsing
- [x] IR
- [x] codgen (doesn't include unit test for codegen)

### CODE REVIEW

**Comment by tohrnii** [comment]
Currently the exponent can only be a constant but in examples written by @bobbinth in the original discussion also allows the exponent to be a variable ([example](https://github.com/maticnetwork/miden/discussions/254#discussioncomment-3293836)). Also, I'm wondering if there is any utility to allow more complex expressions as exponent, for example `col ^ (m + n)`. What do you think @grjte?

File: ir/src/transition_constraints/graph.rs [hunk 1]
```diff
@@ -84,6 +84,18 @@ impl AlgebraicGraph {
...
+            Expr::Exp(lhs, rhs) => {
+                // add base subexpression.
+                let lhs = self.insert_expr(*lhs, trace_columns)?;
+                // add exponent subexpression.
+                if let Expr::Constant(exponent) = *rhs {
+                    Ok(self.insert_op(Operation::Exp(lhs, exponent as usize)))
+                } else {
+                    Err(SemanticError::InvalidExprType(
+                        "Exponent can only be a constant".to_string(),
+                    ))
```
Comment (grjte) [inline ir/src/transition_constraints/graph.rs:96]: I think this error would be more useful if it includes information about the `lhs` and `rhs`.

File: parser/src/ast/mod.rs [hunk 2]
```diff
@@ -48,6 +48,7 @@ pub enum Expr {
     Next(Identifier),
     Add(Box<Expr>, Box<Expr>),
     Subtract(Box<Expr>, Box<Expr>),
+    Exp(Box<Expr>, Box<Expr>),
```
Comment (grjte) [inline parser/src/ast/mod.rs:51]: this is super nitty, but since all of the other variants are full words, I think maybe we should use `Exponent` instead of `Exp`. Either that, or we could switch to `Sub`, `Mul`, `Const`, `Var`, which would be fine too (and maybe better?) Feel free to pick whichever you prefer, as long as we're consistent. (In fact, in the graph in the IR I probably should have used `Const`)

File: parser/src/lexer/mod.rs [hunk 3]
```diff
@@ -81,6 +81,9 @@ pub enum Token {
     #[token("-")]
     Minus,
 
+    #[token("^")]
+    Exp,
+
```
Comment (grjte) [inline parser/src/lexer/mod.rs:86]: nit: same thought about `Exp` vs `Exponent` as I left in the ast

File: codegen-winter/src/air/transition_constraints.rs [hunk 4]
```diff
@@ -83,8 +83,17 @@ impl Codegen for Operation {
...
             }
+            Operation::Exp(l_idx, r_idx) => {
+                let lhs = l_idx.to_string(graph);
+                strings.push(lhs);
+
+                strings.push(String::from("^"));
+
+                let rhs = r_idx.to_string();
+                strings.push(rhs);
+            }
```
Comment (grjte) [inline codegen-winter/src/air/transition_constraints.rs:94]: I have the same thought here that I left in the `mul` PR - I think using `format!` would be better.

File: parser/src/lexer/tests.rs [hunk 5]
```diff
@@ -45,11 +45,16 @@ fn keywords_with_identifiers() {
 
 #[test]
 fn multi_arithmetic_ops() {
-    let source = "enf clk' - clk - 1 = 0";
+    let source = "enf clk'^2 - clk' - clk - 1 = 0";
```
Comment (grjte) [inline parser/src/lexer/tests.rs:48]: Same comment here that I left in the `Mul` PR about modifying this test vs. adding new tests

File: parser/src/parser/grammar.lalrpop [hunk 6]
```diff
@@ -87,8 +87,13 @@ TransitionConstraint: TransitionConstraint = {
...
 Expr: Expr = {
-    <lexpr: Expr> "+" <rexpr: NextExpr> => Expr::Add(Box::new(lexpr), Box::new(rexpr)),
-    <lexpr: Expr> "-" <rexpr: NextExpr> => Expr::Subtract(Box::new(lexpr), Box::new(rexpr)),
+    <lexpr: Expr> "+" <rexpr: ExpExpr> => Expr::Add(Box::new(lexpr), Box::new(rexpr)),
+    <lexpr: Expr> "-" <rexpr: ExpExpr> => Expr::Subtract(Box::new(lexpr), Box::new(rexpr)),
+    ExpExpr
+}
+
+ExpExpr: Expr = {
+    <lexpr: NextExpr> "^" <rexpr: NextExpr> => Expr::Exp(Box::new(lexpr), Box::new(rexpr)),
```
Comment (grjte) [inline parser/src/parser/grammar.lalrpop:96]: I don't think we want to allow a `NextExpr` on the right hand side here. Even when we support variables in the future, we don't want to allow trace columns here. I think this is actually the best place to restrict the rhs to a constant - then you don't have to check it later, since it will be handled during parsing.

File: parser/src/parser/tests/expressions.rs [hunk 7]
```diff
@@ -5,6 +5,33 @@ use super::{
 
 // SECTIONS
 // ================================================================================================
+#[test]
+fn exponentiation() {
+    // the operation must be put into a source section, or parsing will fail
+    let source = "
+    transition_constraints:
+        enf clk'^2 + clk - 1 = 0";
```
Comment (grjte) [inline parser/src/parser/tests/expressions.rs:13]: Same thoughts I left in the `Mul` PR about more rigorous testing of edge cases & different combinations of operations.

**Review by grjte — CHANGES_REQUESTED** [review]
Thanks @tohrnii! I left a few comments that are redundant with the Mul PR, so it makes sense to read that one first. Mostly it's the same theme about now being a good time to introduce more rigorous testing.

Regarding your comment - I agree with restricting the exponent to a constant for now. I think we can add the option for it to be a variable in the future when we add support for variables (since currently, identifiers are always trace columns). Similar thought with the complex expressions - I think this should be added when support for parens is added, but for now the only atoms allowed inside the parens should be constants. In the future when variable support is added, we would then allow atoms that are either identifiers or constants, and we'll then need to do more type-checking in the IR (to tell the difference between allowed identifiers like variables and disallowed ones like trace columns)

**Comment by grjte** [comment]
> Similar thought with the complex expressions - I think this should be added when support for parens is added, but for now the only atoms allowed inside the parens should be constants. In the future when variable support is added, we would then allow atoms that are either identifiers or constants, and we'll then need to do more type-checking in the IR (to tell the difference between allowed identifiers like variables and disallowed ones like trace columns)

Actually, thinking about the complexity of excluding variables now when we'll want to include them again later - I don't think it's worth it. I think even in the parens PR we should just keep the exponent rhs as a constant, and support for complex expressions can be added when variables are added (probably v0.2). What do you think?

File: parser/src/ast/mod.rs (hunk 2 above)
Comment (tohrnii) [reply parser/src/ast/mod.rs]: I agree keeping the concise version seems better. So changed these to `Sub`, `Mul`, `Const`, `Var`, etc.

File: parser/src/lexer/mod.rs (hunk 3 above)
Comment (tohrnii) [reply parser/src/lexer/mod.rs]: Went with `Exp` and changed the rest and `Number` to `Num`.

File: parser/src/parser/grammar.lalrpop (hunk 6 above)
Comment (tohrnii) [reply parser/src/parser/grammar.lalrpop]: Makes sense. Thanks.

File: codegen-winter/src/air/boundary_constraints.rs [hunk 8]
```diff
@@ -55,16 +55,19 @@ trait Codegen {
...
+            Self::Sub(lhs, rhs) => {
                 format!("{} - {}", lhs.to_string(), rhs.to_string())
             }
-            Self::Multiply(lhs, rhs) => {
+            Self::Mul(lhs, rhs) => {
                 format!("{} * {}", lhs.to_string(), rhs.to_string())
             }
+            Self::Exp(lhs, rhs) => {
+                format!("({}).exp({})", lhs.to_string(), rhs)
+            }
```
Comment (grjte) [inline codegen-winter/src/air/boundary_constraints.rs:70]: I think parentheses need to be added to the other expressions above like in the `Mul` PR. These can be revisited at the same time those ones are

**Review by grjte — CHANGES_REQUESTED** [review]
Looks great, thank you @tohrnii! There's just one small issue which I left inline (the same one we just looked at for codegen in the mul PR)

nit: no need to change for this PR, but I would prefer for things like keyword renaming to be in its own `refactor` commit in the future

File: codegen-winter/src/air/boundary_constraints.rs (hunk 8 above)
Comment (grjte) [reply codegen-winter/src/air/boundary_constraints.rs:70]: Actually, since this was a pre-existing issue I'll fix it separately and approve this PR

**Review by grjte — APPROVED** [review]
